### ISSUE TITLE
react UJS not rendering components in tests

### ISSUE THREAD

**Issue by georgemillo** [reactjs/react-rails#439]
When I output a component onto the page using `react_component` everything works fine when I test it manually, but my automated tests won't pass. I initially thought it might be a problem with Poltergeist or PhantomJS, but I switched to Selenium and the tests still fail. When I look at the browser window during a Selenium run, or when I call `save_screenshot` in Poltergeist, I can see that my components aren't being rendered at all; it's as if `ReactDOM.render` (or whatever the method is) never gets called. (But when I open up the gem code and add `console.log` statements inside e.g. `mountComponents` in `react_ujs.js.erb` or within the `render` functions of my own components, I can see that those methods *are* being called - they're just not doing anything.)

I made a minimal app that recreates the problem [here](https://github.com/georgemillo/react-poltergeist-fail-example/blob/master/spec/features/example_spec.rb) - run the test in [spec/features/example_spec.rb](https://github.com/georgemillo/react-poltergeist-fail-example/blob/master/spec/features/example_spec.rb) and you can see it fails, but when you visit the tested page at `root_path` in the browser, you can see that the text *does* appear, i.e. the test should pass.

Am I doing something wrong, or is this a bug?

Here's the Gemfile.lock of that example app, if you're curious. Note that it's using Rails 5.0.0.beta1, and I'm testing all this with MRI Ruby 2.3.0p0.

```
GEM
  remote: https://rubygems.org/
  specs:
    actioncable (5.0.0.beta1)
      actionpack (= 5.0.0.beta1)
      celluloid (~> 0.17.2)
      coffee-rails (~> 4.1.0)
      em-hiredis (~> 0.3.0)
      faye-websocket (~> 0.10.0)
      redis (~> 3.0)
      websocket-driver (~> 0.6.1)
    actionmailer (5.0.0.beta1)
      actionpack (= 5.0.0.beta1)
      actionview (= 5.0.0.beta1)
      activejob (= 5.0.0.beta1)
      mail (~> 2.5, >= 2.5.4)
      rails-dom-testing (~> 1.0, >= 1.0.5)
    actionpack (5.0.0.beta1)
      actionview (= 5.0.0.beta1)
      activesupport (= 5.0.0.beta1)
      rack (~> 2.x)
      rack-test (~> 0.6.3)
      rails-dom-testing (~> 1.0, >= 1.0.5)
      rails-html-sanitizer (~> 1.0, >= 1.0.2)
    actionview (5.0.0.beta1)
      activesupport (= 5.0.0.beta1)
      builder (~> 3.1)
      erubis (~> 2.7.0)
      rails-dom-testing (~> 1.0, >= 1.0.5)
      rails-html-sanitizer (~> 1.0, >= 1.0.2)
    activejob (5.0.0.beta1)
      activesupport (= 5.0.0.beta1)
      globalid (>= 0.3.6)
    activemodel (5.0.0.beta1)
      activesupport (= 5.0.0.beta1)
      builder (~> 3.1)
    activerecord (5.0.0.beta1)
      activemodel (= 5.0.0.beta1)
      activesupport (= 5.0.0.beta1)
      arel (~> 7.0)
    activesupport (5.0.0.beta1)
      concurrent-ruby (~> 1.0)
      i18n (~> 0.7)
      json (~> 1.7, >= 1.7.7)
      method_source
      minitest (~> 5.1)
      tzinfo (~> 1.1)
    arel (7.0.0)
    babel-source (5.8.34)
    babel-transpiler (0.7.0)
      babel-source (>= 4.0, < 6)
      execjs (~> 2.0)
    builder (3.2.2)
    byebug (8.2.1)
    capybara (2.5.0)
      mime-types (>= 1.16)
      nokogiri (>= 1.3.3)
      rack (>= 1.0.0)
      rack-test (>= 0.5.4)
      xpath (~> 2.0)
    celluloid (0.17.2)
      celluloid-essentials
      celluloid-extras
      celluloid-fsm
      celluloid-pool
      celluloid-supervision
      timers (>= 4.1.1)
    celluloid-essentials (0.20.5)
      timers (>= 4.1.1)
    celluloid-extras (0.20.5)
      timers (>= 4.1.1)
    celluloid-fsm (0.20.5)
      timers (>= 4.1.1)
    celluloid-pool (0.20.5)
      timers (>= 4.1.1)
    celluloid-supervision (0.20.5)
      timers (>= 4.1.1)
    cliver (0.3.2)
    coffee-rails (4.1.1)
      coffee-script (>= 2.2.0)
      railties (>= 4.0.0, < 5.1.x)
    coffee-script (2.4.1)
      coffee-script-source
      execjs
    coffee-script-source (1.10.0)
    concurrent-ruby (1.0.0)
    connection_pool (2.2.0)
    diff-lcs (1.2.5)
    em-hiredis (0.3.0)
      eventmachine (~> 1.0)
      hiredis (~> 0.5.0)
    erubis (2.7.0)
    eventmachine (1.0.8)
    execjs (2.6.0)
    faye-websocket (0.10.2)
      eventmachine (>= 0.12.0)
      websocket-driver (>= 0.5.1)
    globalid (0.3.6)
      activesupport (>= 4.1.0)
    hiredis (0.5.2)
    hitimes (1.2.3)
    i18n (0.7.0)
    jquery-rails (4.0.5)
      rails-dom-testing (~> 1.0)
      railties (>= 4.2.0)
      thor (>= 0.14, < 2.0)
    json (1.8.3)
    loofah (2.0.3)
      nokogiri (>= 1.5.9)
    mail (2.6.3)
      mime-types (>= 1.16, < 3)
    method_source (0.8.2)
    mime-types (2.99)
    mini_portile2 (2.0.0)
    minitest (5.8.3)
    multi_json (1.11.2)
    nokogiri (1.6.7.1)
      mini_portile2 (~> 2.0.0.rc2)
    poltergeist (1.8.1)
      capybara (~> 2.1)
      cliver (~> 0.3.1)
      multi_json (~> 1.0)
      websocket-driver (>= 0.2.0)
    puma (2.15.3)
    rack (2.0.0.alpha)
      json
    rack-test (0.6.3)
      rack (>= 1.0)
    rails (5.0.0.beta1)
      actioncable (= 5.0.0.beta1)
      actionmailer (= 5.0.0.beta1)
      actionpack (= 5.0.0.beta1)
      actionview (= 5.0.0.beta1)
      activejob (= 5.0.0.beta1)
      activemodel (= 5.0.0.beta1)
      activerecord (= 5.0.0.beta1)
      activesupport (= 5.0.0.beta1)
      bundler (>= 1.3.0, < 2.0)
      railties (= 5.0.0.beta1)
      sprockets-rails (>= 2.0.0)
    rails-deprecated_sanitizer (1.0.3)
      activesupport (>= 4.2.0.alpha)
    rails-dom-testing (1.0.7)
      activesupport (>= 4.2.0.beta, < 5.0)
      nokogiri (~> 1.6.0)
      rails-deprecated_sanitizer (>= 1.0.1)
    rails-html-sanitizer (1.0.2)
      loofah (~> 2.0)
    railties (5.0.0.beta1)
      actionpack (= 5.0.0.beta1)
      activesupport (= 5.0.0.beta1)
      method_source
      rake (>= 0.8.7)
      thor (>= 0.18.1, < 2.0)
    rake (10.4.2)
    react-rails (1.5.0)
      babel-transpiler (>= 0.7.0)
      coffee-script-source (~> 1.8)
      connection_pool
      execjs
      rails (>= 3.2)
      tilt
    redis (3.2.2)
    rspec-core (3.1.7)
      rspec-support (~> 3.1.0)
    rspec-expectations (3.1.2)
      diff-lcs (>= 1.2.0, < 2.0)
      rspec-support (~> 3.1.0)
    rspec-mocks (3.1.3)
      rspec-support (~> 3.1.0)
    rspec-rails (3.1.0)
      actionpack (>= 3.0)
      activesupport (>= 3.0)
      railties (>= 3.0)
      rspec-core (~> 3.1.0)
      rspec-expectations (~> 3.1.0)
      rspec-mocks (~> 3.1.0)
      rspec-support (~> 3.1.0)
    rspec-support (3.1.2)
    sprockets (3.5.2)
      concurrent-ruby (~> 1.0)
      rack (> 1, < 3)
    sprockets-rails (3.0.0)
      actionpack (>= 4.0)
      activesupport (>= 4.0)
      sprockets (>= 3.0.0)
    sqlite3 (1.3.11)
    thor (0.19.1)
    thread_safe (0.3.5)
    tilt (2.0.2)
    timers (4.1.1)
      hitimes
    tzinfo (1.2.2)
      thread_safe (~> 0.1)
    uglifier (2.7.2)
      execjs (>= 0.3.0)
      json (>= 1.8.0)
    websocket-driver (0.6.3)
      websocket-extensions (>= 0.1.0)
    websocket-extensions (0.1.2)
    xpath (2.0.0)
      nokogiri (~> 1.3)

PLATFORMS
  ruby

DEPENDENCIES
  byebug
  capybara
  jquery-rails
  poltergeist
  puma
  rails (>= 5.0.0.beta1, < 5.1)
  react-rails
  rspec-rails
  sqlite3
  tzinfo-data
  uglifier (>= 1.3.0)

BUNDLED WITH
   1.10.6
```

**Comment by rmosolgo** [comment]
Hmmm what you're doing looks right to me. Here's the only JS test we have in `react-rails`: 

https://github.com/reactjs/react-rails/blob/master/test/react/rails/react_rails_ujs_test.rb

It looks like that uses poltergeist too, but with `inspector: true`. I haven't tried that before, but it _says_ it will open the page in the browser with `page.driver.debug`. Does that work for you? Can you see the components if you do that?

It's so strange that React component functions _are_ called but they don't show up on the page. Are _unmount_ hooks called? (Maybe the component is immediately unmounted for some reason?)

**Comment by georgemillo** [comment]
Whoops, I've just realised that my sample app didn't work properly because I forgot to add `js: true` to the test. After adding `js: true` the spec passes.

But things still fail in my real app, and those tests aren't missing a `js: true`. Hmmmm.... I'll dig deeper and try and recreate the problem in the sample app for real this time.

**Comment by georgemillo** [comment]
Urgh... I'm an idiot. Turns out the problem was with the way I'd configured my JS tests and nothing to do with react-rails. Apologies.

**Comment by rystraum** [comment]
@georgemillo Sorry to ask this about a year later, but I'm running into the same problems. How did you fix your issue?

**Comment by georgemillo** [comment]
@rystraum I can't remember what the problem was with "the way I'd configured my JS tests", sorry.

However I ended up not using `react-rails` at all, and scrapping the asset pipeline altogether in favour of using NPM; it's just not worth the effort otherwise. I've formed the opinion that, if you need something as advanced as React for your front-end, the asset pipeline is more trouble than it's worth and you're better off taking advantage of more modern front-end tools. Not to knock on the hard work of the people who created `react-rails` because I'm sure it has its use cases; just saying that it wasn't right for my purposes.

**Comment by rystraum** [comment]
@georgemillo Thanks for replying. I'm starting to be of the same opinion too. Hopefully with 5.1, it would be less painful to pull in Webpack to Rails projects.

**Comment by cmsepul** [comment]
hey @rystraum , @georgemillo I also started pulling in `react-rails` to get started with React because I thought it was the fastest way, but I found there was a tool to go all in with NPM for this, https://github.com/shakacode/react_on_rails. You might want to check it out if you are thinking on moving away from `react-rails`. If I can't solve this test issue (I got here because I'm having the same issue) I might end up using ReactOnRails earlier than I thought.
